contracts/libraries/SafeMigration.sol

```diff
@@ -0,0 +1,136 @@
+        SAFE_FALLBACK_HANDLER_CODEHASH = getCodehash(fallbackHandler);
+    }
+
+    /**
+     * @notice Migrate the Safe contract to a new Safe singleton implementation.
+     *         Checks whether the given target singleton address codehash matches the expected Safe/SafeL2 singleton codehash.
+     * @param target Address of the new Safe singleton implementation.
+     */
+    function migrateSingleton(address target) public onlyDelegateCall {
+        require(getCodehash(target) == SAFE_SINGLETON_CODEHASH, "Invalid Safe singleton");
+        singleton = target;
+        emit ChangedMasterCopy(target);
+    }
+
+    /**
+     * @notice Migrate to Safe Singleton and set the fallback handler. This function is intended to be used when migrating
+     *         a Safe to a version which also requires updating fallback handler.
+     */
+    function migrateWithFallbackHandler(address target, address fallbackHandler) public onlyDelegateCall {
+        migrateSingleton(target);
+        require(getCodehash(fallbackHandler) == SAFE_FALLBACK_HANDLER_CODEHASH, "Invalid fallbackhandler");
+        ISafe(address(this)).setFallbackHandler(fallbackHandler);
+    }
+
+    /**
+     * @notice Migrate the Safe contract to a new Safe singleton implementation.
+     *         Checks whether the given target singleton address codehash matches the expected Safe/SafeL2 singleton codehash.
+     * @param target Address of the new Safe singleton implementation.
+     */
+    function migrateL2Singleton(address target) public onlyDelegateCall {
+        require(getCodehash(target) == SAFE_L2_SINGLETON_CODEHASH, "Invalid SafeL2 singleton");
+        singleton = target;
+        emit ChangedMasterCopy(target);
+    }
+
+    /**
+     * @notice Migrate to Safe Singleton (L2) and set the fallback handler. This function is intended to be used when migrating
+     *         a Safe to a version which also requires updating fallback handler.
+     */
+    function migrateL2WithFallbackHandler(address target, address fallbackHandler) public onlyDelegateCall {
+        migrateL2Singleton(target);
+        require(getCodehash(fallbackHandler) == SAFE_FALLBACK_HANDLER_CODEHASH, "Invalid fallbackhandler");
+
+        ISafe(address(this)).setFallbackHandler(fallbackHandler);
+    }
+
+    /**
+     * @notice Checks whether an Ethereum address corresponds to a contract or an externally owned account (EOA).
+     * @param account The Ethereum address to be checked.
+     * @return A boolean value indicating whether the address is associated with a contract (true) or an EOA (false).
+     * @dev This function relies on the `extcodesize` assembly opcode to determine whether an address is a contract.
+     * It may return incorrect results in some edge cases (see documentation for details).
+     * Developers should use caution when relying on the results of this function for critical decision-making.
+     */
+    function isContract(address account) internal view returns (bool) {
```

Comment:
naming-nit: I think we should rename this function. Checking if an account is an EOA is not really possible, and this really checks whether or not the account has code (a contract during the constructor has no code, but is not an EOA, precompiles have no code but are also not EOAs, random addresses have no code but are not EOAs, etc.).

```suggestion
    function hasCode(address account) internal view returns (bool) {
```